### PR TITLE
Modernize binaryen.js glue code

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

**Comment by dcodeIO** [comment]
The errors here might be due to renaming performed by CC. For example

```js
  self['local'] = {
    'get': function(index, type) { ... },
```

tells CC not to rename `get` to something shorter during optimizations, while I'd assume that

```js
  self['local'] = {
    get(index, type) { ... },
```

will trigger renaming and as such break.

**Comment by MaxGraey** [comment]
Hmm, not sure about this. [playground works as expected](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520SIMPLE_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540language_in%2520ECMASCRIPT6%250A%252F%252F%2520%2540language_out%2520ECMASCRIPT6%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250Avar%2520test%2520%253D%2520%257B%250A%2520%2520hello(arg)%2520%257B%250A%2520%2520%2520%2520return%2520arg%2520%252B%25201%253B%250A%2520%2520%257D%252C%250A%2520%2520world%253A%2520%257B%250A%2520%2520%2520%2520set(a)%2520%257B%2520return%2520a%2520-%25201%253B%2520%257D%252C%250A%2520%2520%2520%2520get()%2520%257B%2520return%25200%253B%2520%257D%252C%250A%2520%2520%257D%250A%257D%253B%250A%250Awindow%255B'test'%255D%2520%253D%2520test%253B%250A%250A) But emscripten uses half year old CC it seems

**Comment by dcodeIO** [comment]
Hmm, interesting. Do you know if Emscripten uses `ADVANCED_OPTIMIZATIONS`?

**Comment by MaxGraey** [comment]
Hmm. I'm wondering [if this](https://github.com/emscripten-core/emscripten/blob/6d88fbc04aebf19dbc775426b1d51ecd6ba6c903/tools/building.py#L1141) could override your setup for `ECMASCRIPT6`?

**Comment by dcodeIO** [comment]
Hmm, maybe. I don't know enough about Emscripten's integration with CC unfortunately.

**Comment by MaxGraey** [comment]
`Exception: ('run_command failed (7)', 'TypeError: module.i32.atomic.load8_u is not a function\n` even for `'load8_u'(...)`. So it seems we still have CC in ES5 mode =(

**Comment by dcodeIO** [comment]
That points at the atomic variant, but you only added quotes to `i32.load8_u` it seems.

**Comment by MaxGraey** [comment]
Good catch!

**Comment by dcodeIO** [comment]
One way to test if it picks up `ECMASCRIPT6` could be to replace some `var`s with `let`s and/or inline functions with arrow functions. Using these features made the build explode reliable at least the last time I checked. Ofc all of the instructions (or rather, everything that must never become renamed) also needs quotes.

File: src/js/binaryen.js-post.js [hunk 1]
```diff
@@ -524,69 +524,63 @@ function wrapModule(module, self) {
...
-    return preserveStack(function() {
-      return Module['_BinaryenBlock'](module, name ? strToStack(name) : 0,
-                                      i32sToStack(children), children.length,
-                                      typeof type !== 'undefined' ? type : Module['none']);
-    });
+    return preserveStack(() => (
+      Module['_BinaryenBlock'](module, name ? strToStack(name) : 0,
+                               i32sToStack(children), children.length,
+                               typeof type !== 'undefined' ? type : Module['none'])
+    ));
```
Comment (dcodeIO) [inline src/js/binaryen.js-post.js:535]: The additional parenthesis here and at other locations look somewhat odd to me:

```suggestion
    return preserveStack(() =>
      Module['_BinaryenBlock'](module, name ? strToStack(name) : 0,
                               i32sToStack(children), children.length,
                               typeof type !== 'undefined' ? type : Module['none'])
    );
```

Or is there a good reason to do it the other way?

Comment (MaxGraey) [reply src/js/binaryen.js-post.js]: Removed this

**Comment by MaxGraey** [comment]
@kripken It's ready for review

File: src/js/binaryen.js-post.js [hunk 2]
```diff
@@ -4340,34 +4238,34 @@ Module['TupleMake'] = makeExpressionWrapper({
       Module['_BinaryenTupleMakeRemoveOperandAt'](expr, --prevNumOperands);
     }
   },
-  'getOperandAt': function(expr, index) {
+  'getOperandAt'(expr, index) {
```
Comment (kripken) [inline src/js/binaryen.js-post.js:4241]: I wasn't aware this type of syntax is possible (with quotes). Is it common?

Comment (MaxGraey) [reply src/js/binaryen.js-post.js:4241]: It introduced with ES6 / ES2015 so it's not widely known

**Pull request merged**